src/Raven.Server/Documents/Handlers/Processors/Indexes/IndexHandlerProcessorForConvertAutoIndex.cs

```diff
@@ -0,0 +1,716 @@
+
+                    var fieldPath = $"item.{kvp.Key}";
+
+                    sb.AppendLine($"{f.FieldName} = {fieldPath},");
+                }
+            }
+        }
+    }
+
+    private static void ConstructReduceInternal(StringBuilder sb, AutoIndexDefinition autoIndex)
+    {
+        sb
+            .AppendLine("from result in results")
+            .AppendLine($"group result by {GenerateGroupBy(autoIndex)} into g")
+            .AppendLine("select new")
+            .AppendLine("{");
+
+        foreach (var kvp in autoIndex.MapFields)
+        {
+            var fieldNames = GenerateFieldName(kvp.Key, kvp.Value.Indexing);
+
+            foreach (var f in fieldNames)
+            {
+                switch (kvp.Value.Aggregation)
+                {
+                    case AggregationOperation.Count:
+                        {
+                            var fieldPath = $"g.Sum(x => x.{kvp.Key})";
+
+                            sb.AppendLine($"{f.FieldName} = {fieldPath},");
+                            break;
+                        }
+                    case AggregationOperation.Sum:
+                        {
+                            var fieldPath = $"g.Sum(x => x.{kvp.Key})";
+
+                            sb.AppendLine($"{f.FieldName} = {fieldPath},");
+                            break;
+                        }
+                    default:
+                        throw new NotSupportedException($"Field '{kvp.Key}' has unsupported aggregation type '{kvp.Value.Aggregation}'.");
+                }
+            }
+        }
+
+        foreach (var kvp in autoIndex.GroupByFields)
+        {
+            var groupByField = GenerateFieldName(kvp.Key, AutoFieldIndexing.Default).Single();
+            var fieldNames = GenerateFieldName(kvp.Key, kvp.Value.Indexing);
+
+            foreach (var f in fieldNames)
+            {
+                if (f.IsConstant)
+                    continue;
+
+                var fieldPath = $"g.Key.{groupByField.FieldName}";
+
+                sb.AppendLine($"{f.FieldName} = {fieldPath},");
+            }
+        }
+
+        sb.AppendLine("};");
+
+        return;
+
+        static string GenerateGroupBy(AutoIndexDefinition autoIndex)
+        {
+            var seenConst = false;
+            var sb = new StringBuilder();
+            foreach (var fieldName in autoIndex.GroupByFieldNames)
+            {
+                if (sb.Length > 0)
+                    sb.Append(", ");
+
+                var f = GenerateFieldName(fieldName, indexing: null).Single();
+                if (f.IsConstant)
+                {
+                    sb.Append(f.FieldName);
+                    seenConst = true;
+                }
+                else
+                {
+                    if (seenConst)
+                        throw new InvalidOperationException("Cannot generate complex group by because const was encountered.");
```

Comment:
What do you think about rephrasing it to "because a constant field was encountered."?